### ISSUE TITLE
Creating a monodimensional sparse array returns an error

### ISSUE THREAD

**Issue by marfago** [apache/incubator-mxnet#8817]
Hello,

The following code
```
b1 = nd.sparse.zeros('csr',shape=10, ctx=ctx)
```

returns

```
/usr/local/lib/python3.6/site-packages/mxnet/ndarray/sparse.py in _new_alloc_handle(stype, shape, ctx, delay_alloc, dtype, aux_types, aux_shapes)
     87     check_call(_LIB.MXNDArrayCreateSparseEx(
     88         ctypes.c_int(int(_STORAGE_TYPE_STR_TO_ID[stype])),
---> 89         c_array(mx_uint, shape),
     90         mx_uint(len(shape)),
     91         ctypes.c_int(ctx.device_typeid),

/usr/local/lib/python3.6/site-packages/mxnet/base.py in c_array(ctype, values)
    214     2.0
    215     """
--> 216     return (ctype * len(values))(*values)
    217 
    218 def ctypes2buffer(cptr, length):

TypeError: object of type 'int' has no len()
```

but 
```
b1 = nd.zeros(shape=10, ctx=ctx)
```
works fine.

**Comment by piiswrong** [comment]
As the name `Compressed Sparse Row matrix` suggests, it's an matrix, thus must have 2 dimensions

**Comment by marfago** [comment]
Thank you.
Maybe the documentation is just not clear enough, it just says:"Return a new array of given shape and type, filled with zeros."

**Comment by marfago** [comment]
However, I tried also with 
```b1 = nd.zeros(shape=10, ctx=ctx, stype='row_sparse')```
and I got the same error. 

What is the way to build a sparse vector?

**Comment by weixingzhang** [comment]
CSR or row_sparse is for matrix. Thus, the implementation is expecting that
shape is a list rather than an integer.

The code below should work.

b1 = mx.nd.sparse.zeros('csr',shape=[10,1], ctx=ctx)

**Comment by marfago** [comment]
The @wangxianliang 's code assumes that a vector and a 2D matrix with only one column are equivalent (and actually they are). If this holds always for the current implementation, why does the nd.sparse.zeros function fails instead of silently converting shape=(10) to shape(10,1)? It will keep nd.sparse.zeros and nd.sparse interchangeable.

**Comment by eric-haibin-lin** [comment]
Thanks for reporting. I think we can add additional logic to handle such case. I tried scipy.sparse as follows:
```
>>> np.ones((10))
array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
>>> a = np.ones((10))
>>> a.shape
(10,)
>>> import scipy.sparse as spsp
>>> spsp.csr_matrix(a)
<1x10 sparse matrix of type '<type 'numpy.float64'>'
        with 10 stored elements in Compressed Sparse Row format>
```
Instead of converting (10) to (10,1), I think (1,10) will be more appropriate for CSR since it expects large number of columns.

**Comment by weixingzhang** [comment]
@eric-haibin-lin I just made a change for this issue. Can you help review it? Thanks! [https://github.com/apache/incubator-mxnet/pull/8843](url)

**Comment by eric-haibin-lin** [comment]
@reminisce @anirudh2290 @ZiyueHuang @piiswrong  any comment on converting shape (10) to (1,10) explicitly for csr & row_sparse arrays? scipy always performs that:
```
>>> a = np.ones((10))
>>> a.shape
(10,)
>>> import scipy.sparse as spsp
>>> spsp.csr_matrix(a)
<1x10 sparse matrix of type '<type 'numpy.float64'>'
        with 10 stored elements in Compressed Sparse Row format>
>>> spsp.csc_matrix(a)
<1x10 sparse matrix of type '<type 'numpy.float64'>'
        with 10 stored elements in Compressed Sparse Column format>
```

**Comment by reminisce** [comment]
I think it's natural to reshape a vector of shape (10,) to a matrix of shape (1, 10), placing 10 elements in the same row by inserting an axis before the existing one, so I agree that we follow the scipy's convention.

**Comment by ZiyueHuang** [comment]
LGTM

**Comment by anirudh2290** [comment]
+1 for converting shape (10) to (1, 10) for CSR. IMO, this behavior should be consistent between row_sparse and csr (in both cases (10) should convert to (1, 10)).

**Comment by marfago** [comment]
In my opinion the behaviour should be consistent between 

mx.nd.zeros(10, mx.cpu())  -> shape=(10,)
mx.nd.zeros(10, mx.cpu(), stype='csr') -> error
mx.nd.zeros((1,10), mx.cpu(), stype='csr') -> shape(1,10)

so that a csr vector or a dense vector can be used in the same code without any particular adaptation.
It also requires some other changes around in the code (#8883)

**Comment by weixingzhang** [comment]
I agree to convert (10) to (1, 10) and make the usages below consistent.

mx.nd.zeros(10, mx.cpu()) -> shape=(10,)
mx.nd.zeros(10, mx.cpu(), stype='csr') -> error
mx.nd.zeros((1,10), mx.cpu(), stype='csr') -> shape(1,10)

**Comment by eric-haibin-lin** [comment]
From an implementation point of view, it's hard to enable such mono-dimensional array. 

The 1-D shape could be passed directly to the backend, whereas all the sparse operators in the backend  assumes `shape[0]` is num_rows and `shape[1]` is num_cols at the moment. Accepting 1-D shape for csr would require **significant efforts** to check these cases for each operator in the backend. `zeros` is not the only operator affected, cases include:
- `mx.sym.var('a', shape=10, stype='csr')`
- generate a random uniform csr ndarray with 1-D shape 
- dot with 1-D csr
- convert a dense 1-D array to csr
- etc..

I suggest that we throw an exception for 1-D shape for `sparse.zeros` to guide users the expected usage instead.
